README.md
# challenge-toolbox

The challenge toolbox is a small component which allows you to create, run and check new avatao challenges locally. If you have any questions don't hesitate to contact us at [content@avatao.com](mailto:content@avatao.com).

## Rules of participation

[Avatao](https://avatao.com) is a company that wants to make hands-on training accessible to everyone. This includes open-source communities and other businesses also.

- **Ownership of challenges**: We released this tool to help people to run their own exercises locally and share with others.
- **Open-source licensing**: The value to the community contribution is huge - each one is given in service and respect to the community. To respect this, we release the challenge toolbox and the related templates under *Apache 2.0* License. We do not restrict the use of derived challenges as long as they are contributed back under the same license.
- **Business services**: Avatao provides a 24/7 runtime environment that allows you to expose your challenge online in an easy way. Note that Avatao provides additional services (e.g., custom-tailored challenges for businesses, community organizations, enterprise support, training programs, customized learning) as part of a business offering.


## Create your own challenge

### Video tutorial \[new\]

We recorded a [tutorial video series](https://www.youtube.com/watch?v=wlaEQlXo8N8&list=PLALqwhWHQH2aryTOTrzfU49hvmB0aNxRk) to help you create your own avatao security challenges.

### Prerequisites

* Docker
* Python 3
* Run `pip3 install -r requirements.txt`


### Quick reference

* Choose a challenge template from the `templates` directory
* Build challenge: `./build.py <challenge_folder>` (e.g., `python3 ./build.py templates/xss`)
* Start challenge: `./start.py <challenge_folder>` (e.g., `python3 ./start.py templates/xss`)
* Check challenge format: `./check-format.py <challenge_folder>` (e.g., `python3 ./check-format.py templates/xss`)
* Check challenge solution `./check-solution.py <optional_flag>` (e.g., `python3 ./check-solution.py flag`)
* Create tests to check the correct solution
* Cleanup: `./docker-cleanup.sh`


### **Start with a challenge template**

So as to ease challenge development we've prepared templates for different challenge types using our [base images](https://hub.docker.com/u/avatao/) (e.g., ubuntu, controller, exploitation, web-ide). These templates contain examples for static and container-based challenges. Please use the one which fit your needs and customize it as you like.

- [Static challenges without containers](templates/file)
- [Challenges running a server and accept client connections via telnet](templates/telnet)
- [Challenges accepting connections via SSH](templates/ssh)
- [C programming challenges](templates/c)
- [C# programming challenges](templates/csharp)
- [Java programming challenges](templates/java)
- [XSS challenges](templates/xss)

By cloning this repository you get **scripts** and **skeleton** files that will help you a lot in challenge creation.

### **Build**

To build your challenge images (e.g., solvable, controller) please use our
build script as follows.

    ./build.py <repository_path>

![build](img/build.gif)

For example, if your repository path is `/home/user/my-challenge`, your images will be called
`my-challenge-solvable` and `my-challenge-controller`, if you have `Dockerfile` for both controller and solvable.

If you do not want to build both images, just simply modify the name of image (e.g., `solvable_`) you do not need now and the script skips it.

Only the first build is slow. If you modify any file in those directories, you need to run a rebuild before starting them.

If the build process was successful you should see your new docker image by typing the following command:

    docker images


### **Start**

To start your challenge, simply type:

    ./start.py <repository_path>

![start](img/start.gif)

When a controller-solvable pair is started, you can address them internally as `localhost`, thus no IP address is required. This is useful, for example, when you want to access the solvable with a solution checking script from the controller container. See the [example](templates/xss/controller/opt/check.js) for more details.

![browser](img/browser.png)


### **Modify and check format**

In the next step, please modify the cloned template in accordance with
your challenge. If you have prepared everything, run our checker script of this guide repository by simply
typing.

    ./check-fomat.py <repository_path>

For example, if your challenge is located at path `/home/user/my-challenge`, your command is the following.

    ./check-format.py /home/user/my-challenge

![check](img/check.png)


Note that `check-format.py` will fail until you start your challenge, however, it is good to verify if everything is in place.
Please fine-tune your files until you see errors. This script will help you a lot:

1. Checks if your files are in place
2. Checks if your configuration (config.yml) and markdown files (e.g., writeup.md) are correct
3. Invokes solution check and the **test** function automatically for docker-based challenges to check if your challenge is working well.


### **Manual solution checking**

Avatao platform calls into the controller via HTTP with the (optional) flag that user submitted. You will be able test that functionality locally with following command:

    ./check-solution.py <optional_flag>



### **Manual challenge testing**

Before every challenge deployment on avatao, we automatically test if the challenge is working properly. To do that, you need to implement the `test` function in the controller's `server.py`. You can find [here](templates/c/controller/opt/server.py) an example `test` function implementation for C programming challenges.

After implementing the `test` function you can simply run `curl` to make sure that everything works well.

    curl 127.0.0.1:5555/secret/test

Similarly to solution checking, the `test` function returns with HTTP 200 status code if all the tests passed and
returns with the build log. Otherwise the HTTP status code is 500.


### **Clean**

Sometimes it is worth removing stalled and dangling images. To do that simply
run our clean-up script:

    ./docker-cleanup.sh



## Best practices

We have prepared many useful packages and dependencies in our base images and templates, but you might need to get your hands dirty and add extra stuff. To help you, we summarized the most important best practices you should know about docker.

### Docker commands and options

- If you need to install extra packages to controller use the following template:

        FROM avatao/controller:ubuntu-14.04

        USER root
        RUN apt-get -qy update \
            && apt-get install -qy <your_package>
        USER user

        COPY ./controller/ /


- The `USER user` docker command should be used in the solvable `Dockerfile` if a service can run without root privileges.
- Please, make sure that none of the user-controllable services run with root privileges. For example, an SSH daemon can run as root but it should only allow non-privileged users to log in.
- The `COPY` docker command is the preferred way to add your files directories to a container. It has various advantages over `ADD`: 1) simple, 2) preserves file attributes, 3) do not decompress archives or fetch URLs.
- Minimize filesystem layers by reducing the number of `RUN` commands.
- If you do not delete something in the same `RUN` command it was created in, it will still be part of the final image in an internal layer.
- The `find /bin /sbin /usr -perm /6000 -exec chmod -s '{}' \;` command disables suid flags on system files so we could safely keep the `SETGID` and `SETUID` capabilities for a challenge.

There are various runtime docker options that we plan to turn on when starting your challenge with the `docker run` command.

- The `--read-only` option mounts the container's root filesystem as read only. As certain challenges require to have writable directories and files, you can add volumes with the `VOLUME` docker command. For example we added the following volumes for LAMP base images by default.
    ```
    VOLUME ["/etc/mysql", "/var/lib/mysql", "/etc/php5/apache2", "/var/log", "/run"]
    ```
- By default we drop all the capabilities with `--cap-drop=ALL` as you can see in `start.py`. To add extra capabilities please use the capabilities field of `config.py`. If you do not need extra capability just simply leave the array empty (['']).

_Reference_:

* https://docs.docker.com/engine/reference/builder/
* https://docs.docker.com/engine/reference/commandline/run/
* https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities
* https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/

### **Challenge debugging**

It might take too much time to keep waiting for `build.py` and `start.py` to finish when you want to debug your challenge.
It is much simpler to start your challenges with `start.py` and the run

    docker exec -ti <name> bash

![debug](img/debug.png)

For example, if you have problems with implementing the test endpoint in the controller, it is better if you simply
copy `server.py` at a writeable path (e.g., added with the `VOLUME` command) and edit locally with `vim` for example.
**IMPORTANT** Don't forget to save your changes outside the container regularly, as all your modifications will be lost
if you stop `start.py`.


### **Random flag generation**

In order to avoid trivial user cheats flags can be dynamically generated every time a challenge is started. Here are two options:
- **flag = static part + random part**: In this case, a fix static part is extended with a suffix randomly generated whenever the challenge is started.

- **totally random flag**: In the code below, we store an entirely random flag in the solvable's `/var/flag/flag.txt` file. This file is also accessible for the controller container.

    controller/opt/server.py
    ```
    #random flag generation
    with open("/home/user/flag.txt","w") as f:
        f.write("Ea5Y_Pea5Y_")
        for x in random.sample(range(1, 5000), 8):
            f.write(str(x))
            if x%2 == 0:
                f.write("_/")

    # solution checking
    def solution_check():
        submitted_solution = request.json['solution']
        if submitted_solution != open('/var/flag/flag.txt').read():
            return jsonify(solved=False)
        return jsonify(solved=True)
    ```

### **Writeup guide**

1. Every challenge should have an own separate `writeup.md` file under the `metadata` directory.
2. Add the challenge name in H1 style as the example [writeup.md](skeleton/metadata/writeup.md) shows it.
3. The writeup should have at least 3 sections.
4. The cost of first section should be 0%, because it's just a detailed "What to do here?".
5. Each section describes a relevant part from the complete solution. Thus, users should be able to solve the challenge by simply following the instructions of each section.
6. Every section begins with "## " which is an atx-style header (H2) and continues with a name which summarizes well the contents of the section (e.g., contains specific keywords). Take care of inserting a space after "##" to be compatible with the markdown standard.
7. When a user requests hints we first show him the section names mentioned above with the cost of **10%** of the challenge score. We suppose that the name of sections helps users enough to continue the challenge or at least they can choose which section to open. In this way, we only deduct the cost of requested sections and this additional **10%** from the challenge score.
8. Under each section the "Cost: x%" tag shows the percentage that we deduct from the challenge score if a user asks for this part of the solution. The sum of costs of all but the last sections should be equal to **90%**!
9. The last section should be called `Complete solution` which does not have a cost as it is only revealed when somebody asks for the complete solution. In that case we show him the entire writeup.
*IMPORTANT* We highly recommend to insert a few lines of notes about mitigations (e.g., how to fix the vulnerability) if the challenge is an offensive one.

## Documentation

[Please read the docs for more information](docs/README.md).

## Troubleshooting

### *I'm getting port already taken errors when starting a solvable-controller pair*
Please don't try to run more than one solvable-controller pairs.
If it is not the case, something else went wrong and kill all docker containers using:

    docker kill $(docker ps -q)

If you get docker usage message upon killing, then there were no running containers.

### *Solvable starts, but connection attempts fail with:*

    ssh_exchange_identification: Connection closed by remote host

You run a command in solvable which blocks the start of SSH daemon. Fix the problematic command, and rebuild. This weird behavior is observed because docker accepts connection attempts to published ports but then resets them if the published port isn't really listening.

### *I need to enter the container for error checking:*
Just build the container with `build.py`, check the built image with `docker image` and :

    docker run -ti <repository|ID> bash


CHANGES.md
# Changes from v2 to v3

* `type` has been renamed to `crp_type` and its specification changed...
* If `crp_type` is not set then the challenge is static (no containers or VMs).
* Supported `crp_type` values: `docker` same as before, `gce` for new VM based challenges on GCE.
* `crp_config` is different for container vs VM based providers.
* `crp_config.<container>.mem_limit` has been converted to `crp_config.<container>.mem_limit_mb`.
* All metadata fields and files will be handled by the new platform instead.


check-format.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# -*- mode: python; -*-
import json
import logging
import os
import re
import subprocess
from glob import glob as glob

from common import get_sys_args, read_config, yield_dockerfiles
from common import init_logger, counted_error, at_exit

CAPABILITIES = {
    # all linux capabilities:
    'SETPCAP', 'SYS_MODULE', 'SYS_RAWIO', 'SYS_PACCT', 'SYS_ADMIN', 'SYS_NICE',
    'SYS_RESOURCE', 'SYS_TIME', 'SYS_TTY_CONFIG', 'MKNOD', 'AUDIT_WRITE',
    'AUDIT_CONTROL', 'MAC_OVERRIDE', 'MAC_ADMIN', 'NET_ADMIN', 'SYSLOG', 'CHOWN',
    'NET_RAW', 'DAC_OVERRIDE', 'FOWNER', 'DAC_READ_SEARCH', 'FSETID', 'KILL',
    'SETGID', 'SETUID', 'LINUX_IMMUTABLE', 'NET_BIND_SERVICE', 'NET_BROADCAST',
    'IPC_LOCK', 'IPC_OWNER', 'SYS_CHROOT', 'SYS_PTRACE', 'SYS_BOOT', 'LEASE',
    'SETFCAP', 'WAKE_ALARM', 'BLOCK_SUSPEND',
} - {  # blacklisted capabilities (subject to change):
    'MAC_ADMIN', 'MAC_OVERRIDE', 'SYS_ADMIN', 'SYS_MODULE', 'SYS_RESOURCE',
    'LINUX_IMMUTABLE', 'SYS_BOOT', 'BLOCK_SUSPEND', 'WAKE_ALARM',
}

# Allowed kernel parameters to set via the --sysctl docker run option:
# This may be extended manually on a case-by-case basis.
KERNEL_PARAMETERS = {
    "net.ipv4.ip_forward",
}

PORT_RANGE = {'min': 1, 'max': 65535}

CONTROLLER_PROTOCOL = 'controller'
PROTOCOLS = {'udp', 'tcp', 'ssh', 'http', 'ws', CONTROLLER_PROTOCOL}

DIFFICULTY_RANGE = {'min': 10, 'max': 500}
NAME_RANGE = {'min': 3, 'max': 200}
SUMMARY_RANGE = {'min': 30, 'max': 200}
DESCRIPTION_RANGE = {'min': 100, 'max': 30000}

CONFIG_KEYS = {
    'crp_config', 'difficulty', 'enable_flag_input', 'flag', 'name', 'owners',
    'skills', 'recommendations', 'version',
}

COST_SUM = 90
MIN_WRITEUP_SECTIONS = 3

FORWARD_ADDR = '127.0.0.1'
CONTROLLER_PORT = 5555

TOOLBOX_PATH = os.path.dirname(os.path.realpath(__file__))


def check_config(config: dict, is_static):
    invalid_keys = set(config.keys()) - set(CONFIG_KEYS)
    if len(invalid_keys) > 0:
        counted_error('Invalid key(s) found in config.yml: %s' % invalid_keys)

    if config['version'][:1] != 'v':
        counted_error('Invalid version. The version number must start with the letter v')
    elif config['version'] == 'v1':
        counted_error('This version is deprecated, please use v2.0.0')
    elif config['version'] != 'v2.0.0':
        counted_error('Invalid version. The supplied config version is not supported')

    # Difficulty
    try:
        assert DIFFICULTY_RANGE['min'] <= int(config['difficulty']) <= DIFFICULTY_RANGE['max']
    except Exception:
        counted_error('Invalid difficulty in config.yml. '
                      'Valid values: %d - %d' % (DIFFICULTY_RANGE['min'], DIFFICULTY_RANGE['max']))

    # Name
    try:
        assert NAME_RANGE['min'] <= len(config['name']) <= NAME_RANGE['max']
    except Exception:
        counted_error('Invalid challenge name in config.yml. '
                      'Name should be a string between %d - %d characters.' % (NAME_RANGE['min'], NAME_RANGE['max']))

    for template_config in glob(os.path.join(TOOLBOX_PATH, 'templates', '*', 'config.yml')):
        if config['name'] == read_config(template_config).get('name'):
            counted_error('Please, set the challenge name in the config file.')

    # Skills
    if not isinstance(config['skills'], list):
        counted_error('Invalid skills in config.yml. Skills should be placed into a list.\n'
                      '\tValid skills are listed here: \n'
                      '\thttps://platform.avatao.com/api-explorer/#/api/core/skills/')

    # Recommendations
    if not isinstance(config['recommendations'], dict):
        counted_error('Invalid recommendations in config.yml. Recommendations should be added in the following '
                      'format:\n\n'
                      'recommendations:\n'
                      '\twww.example.com: \'Example webpage\'\n'
                      '\thttp://www.example2.com: \'Example2 webpage\''
                      '\thttp://example3.com: \'Example3 webpage\'')

    # https://mathiasbynens.be/demo/url-regex - https://gist.github.com/dperini/729294
    url_strip_chr = r'!"#$%&\'()*+,-./@:;<=>[\\]^_`{|}~'
    url_re = re.compile(r'^(?:(?:https?|ftp)://)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})'
                        r'(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})'
                        r'(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}'
                        r'(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|'
                        r'(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*'
                        r'(?:\.(?:[a-z\u00a1-\uffff]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$')

    for item in config['recommendations'].items():
        if url_re.fullmatch(item[0].strip(url_strip_chr)) is None:
            counted_error('Invalid recommended URL (%s) found in config.yml' % item[0])

        if not isinstance(item[1], str):
            counted_error('The name of recommended url (%s) should be a string in config.yml' % item[1])

    # Owners
    if not isinstance(config['owners'], list):
        counted_error('Challenge owners (%s) should be placed into a list in config.yml' % config['owners'])

    email_re = re.compile(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)")
    for owner in config.get('owners', []):
        if email_re.fullmatch(owner) is None:
            counted_error('Invalid owner email (%s) found in config.yml. '
                          'Make sure you list the email addresses of the owners.' % owner)

    if not is_static:
        controller_found = False
        for item in config['crp_config'].values():

            if not isinstance(item, dict):
                counted_error('Items of crp_config must be dictionaries.')

            if 'image' in item and item['image'].find('/') < 0:
                counted_error('If the image is explicitly defined, it must be relative '
                              'to the registry - e.g. challenge:solvable.')

            if 'capabilities' in item:
                invalid_caps = set(item['capabilities']) - CAPABILITIES
                if len(invalid_caps) > 0:
                    counted_error('Forbidden capabilities: %s\n\tAllowed capabilities: %s',
                                  invalid_caps, CAPABILITIES)

            if 'kernel_params' in item:
                invalid_parameters = set(item['kernel_params']) - KERNEL_PARAMETERS
                if len(invalid_parameters) > 0:
                    counted_error('Forbidden kernel parameters: %s\n\tAllowed parameters: %s',
                                  invalid_parameters, KERNEL_PARAMETERS)

            if 'mem_limit' in item:

                if not isinstance(item['mem_limit'], str):
                    counted_error('Invalid mem_limit value: %s, The mem_limit should be a string like: 100M')

                if item['mem_limit'][-1] not in "M":
                    counted_error('Invalid mem_limit value: %s, The mem_limit should be a string ending with '
                                  'M (megabytes). No other unit is allowed.')
                try:
                    mem_limit_number_part = int(item['mem_limit'][:-1])
                    if mem_limit_number_part > 999:
                        counted_error('Invalid mem_limit value: %s, The mem_limit can not be greater than 999M.')
                except Exception:
                    counted_error('Invalid mem_limit value: %s, mem_limit must start with a number and end with '
                                  'M (megabytes). No other unit is allowed.')

            for port in item.get('ports', []):
                try:
                    port, protocol = port.split('/', 1)
                    try:
                        port = int(port)
                    except Exception:
                        counted_error('Invalid port. The port should be a number between 1 and 65535.')

                    if PORT_RANGE['min'] > port or PORT_RANGE['max'] < port:
                        counted_error('Invalid port. The port should be a number between 1 and 65535')

                    if protocol not in PROTOCOLS:
                        counted_error('Invalid protocol in config.yml (crp_config): %s. Valid values: %s', protocol,
                                      PROTOCOLS)

                    elif protocol == CONTROLLER_PROTOCOL:
                        controller_found = True

                except Exception:
                    counted_error('Invalid port format. [port/protocol]')

        if not config.get('flag') and not controller_found:
            counted_error('Missing controller port [5555/%s] for a dynamic challenge.' % CONTROLLER_PROTOCOL)

    if str(config.get('enable_flag_input')).lower() not in ('true', 'false', '1', '0'):
        counted_error('Invalid enable_flag_input. Should be a boolean.')

    if is_static:
        try:
            assert isinstance(config['flag'], str)
        except AssertionError:
            counted_error('Invalid flag. Should be a string.')
        except KeyError:
            counted_error('Missing flag. Static challenges must have a static flag set.')

    if 'flag' in config and config['flag'][0:6] == 'regex:':
        try:
            re.compile(config['flag'][6:])
        except Exception:
            counted_error('Failed to compile regex flag.')


def check_dockerfile(filename):
    repo_pattern = 'FROM ((docker\.io\/)?avatao|eu\.gcr\.io\/avatao-challengestore)\/'

    try:
        with open(filename, 'r') as f:
            d = f.read()
            if re.search(repo_pattern, d) is None:
                counted_error('Please use avatao base images for your challenges. Our base images'
                              ' are available at https://hub.docker.com/u/avatao/')
    except FileNotFoundError as e:
        counted_error('Could not open %s' % e.filename)

    except Exception as e:
        counted_error('An error occurred while loading %s. \n\tDetails: %s' % (filename, e))


def check_controller():
    check_dockerfile('controller/Dockerfile')
    static_flag = None

    try:
        static_flag = read_config()['flag']

    except KeyError:
        logging.info('[This is not an error] The flag is missing from the config file.\n\tYou should implement '
                     'a dynamic solution checker (e.g., random flags, unit tests) in the controller.')

    except FileNotFoundError as e:
        counted_error('Could not open %s' % e.filename)

    try:
        with open('controller/opt/server.py') as server_file:
            server = server_file.read()

    except FileNotFoundError as e:
        counted_error('Could not open %s' % e.filename)

    except Exception as e:
        counted_error('An error occurred while checking controller/opt/server.py. \n\tDetails: %s' % e)

    else:
        # Invoke test endpoint if not static
        _http_request('http://%s:%d/secret/test' % (FORWARD_ADDR, CONTROLLER_PORT))

        # Check controller's solution_check endpoint.
        solution_check_pattern = 'def solution_check\(\):'
        solution_check = re.findall(solution_check_pattern, server)

        if not (solution_check or static_flag):
            logging.warning('Function "solution_check()" is missing from controller/opt/server.py. '
                          '\n\tPlease implement it and check user solutions dynamically (e.g., random flag '
                          'checking)\n\tor insert a static flag into config.yml.')


def check_misc():
    if not len(glob('src/*')):
        logging.warning('Missing or empty "src" directory. Please place your source files there '
                        'if your challenge has any. ')

    if not len(glob('README.md')):
        logging.warning('No README.md file is found. Readmes help others to understand your challenge.')

    if not len(glob('LICENSE')):
        logging.warning('No LICENSE file is found. Please add the (original) license file if you copied'
                        '\n\t  a part of your challenge from a licensed challenge.')

    if not len(glob('CHANGELOG')):
        logging.warning('No CHANGELOG file is found. If you modified an existing licensed challenge,\n\t'
                        'please, summarize what your changes were.')
        
    if not len(glob('.drone.yml')):
        logging.warning('No .drone.yml file is found. This file is necessary for our automated tests,\n\t'
                        'please, get it from any template before uploading your challenge.')


def check_yml(filename, is_static: bool=False):
    try:
        config = read_config(filename)
        check_config(config, is_static)

    except FileNotFoundError as e:
        counted_error('Could not open %s' % e.filename)

    except KeyError as e:
        counted_error('Key "%s" is missing from %s' % (e.args[0], filename))

    except Exception as e:
        counted_error('An error occurred while loading %s. \n\tDetails: %s' % (filename, e))


def check_metadata():
    # Check if metadata files exist
    try:
        with open('metadata/description.md', 'r', encoding='utf-8') as description_file:
            _check_description(description_file)

        with open('metadata/summary.md', 'r', encoding='utf-8') as summary_file:
            s = summary_file.read()
            if not SUMMARY_RANGE['min'] <= len(s) <= SUMMARY_RANGE['max']:
                counted_error('Summary should be minimally %d, maximally %d characters long.',
                              SUMMARY_RANGE['min'], SUMMARY_RANGE['max'])

        if os.path.exists('metadata/writeup.md'):
            with open('metadata/writeup.md', 'r', encoding='utf-8') as writeup_file:
                _check_writeup(writeup_file)

    except FileNotFoundError as e:
        counted_error('Missing %s from metadata.' % e.filename)

    except UnicodeDecodeError as e:
        counted_error('Could decode markdown text: \n\n%s \n\tDetails: %s' % (e.object, e))

    except Exception as e:
        counted_error('Could not read file in metadata. \n\tDetails: %s' % e)


def sanity_check():
    # Check if the challenge is static
    is_static = len(list(yield_dockerfiles(repo_path, repo_name))) == 0
    if is_static and len(glob('downloads/*')) == 0:
        logging.warning('Static challenges should have a "downloads" directory for sharing challenge files with users.')

    check_yml('config.yml', is_static)

    if not is_static:
        check_controller()
        check_dockerfile('solvable/Dockerfile')

    check_metadata()
    check_misc()

def filter_source(ignore,pattern,text):
    match = [m.span() for m in re.finditer(pattern,text)]
    for i in ignore:
        match = filter(lambda m: m[0] < i[0] or m[1] > i[1],match)
    return list(match)

def _check_description(description_file):
    description = description_file.read()
    if not DESCRIPTION_RANGE['min'] <= len(description) <= DESCRIPTION_RANGE['max']:
        counted_error('Description should be minimum %d and maximum %d characters long.',
                      DESCRIPTION_RANGE['min'], DESCRIPTION_RANGE['max'])

    section_too_long_pattern = re.compile(r'^ {0,3}#{1,6} +.{151,}$', re.MULTILINE)
    section_too_short_pattern = re.compile(r'^ {0,3}#{1,6} +.{0,4}$', re.MULTILINE)

    section_long = re.findall(section_too_long_pattern, description)
    section_short = re.findall(section_too_short_pattern, description)

    if section_short:
        counted_error('Sections must be between 5 and 150 characters.\n'
                      '\tDetails: %s' % section_short)
    if section_long:
        counted_error('Sections must be between 5 and 150 characters.\n'
                      '\tDetails: %s' % section_long)
    
    tick = zip(*[re.finditer("```",description)]*2)
    ignore = [((p[0].span()[1],p[1].span()[0])) for p in tick]

    section_not_capitalized_pattern = re.compile(r'^ {0,3}#{1,6} +[^A-Z].+.*$', re.MULTILINE)

    section_not_capitalized = filter_source(ignore,section_not_capitalized_pattern, description)

    if section_not_capitalized:
        counted_error('Sections must start with a capitalized letter.\n'
                      '\tDetails: %s' % section_not_capitalized)

    # Description can contain only contain H4- or H5-style sections.
    maxh3_pattern = re.compile(r'^ {0,3}#{1,3} +.*$', re.MULTILINE)
    maxh3 = filter_source(ignore,maxh3_pattern, description)
    if maxh3:
        counted_error('A section name font size is too large in description.md. '
                      'Please, use H4 (####) or H5 (#####) section names.\n'
                      '\tDetails: %s' % maxh3)


def _check_writeup(writeup_file):
    writeup = writeup_file.read()

    # Check if writeup.md starts with the challenge name in H1 style.
    config = read_config('config.yml')
    config_challenge_name = config['name']

    h1_pattern = '^%s\n={1,}\n\n' % re.escape(config_challenge_name)
    try:
        writeup_challenge_name = re.search('^.*\n', writeup).group()[:-1]
        assert writeup_challenge_name == config_challenge_name
        re.search(h1_pattern, writeup).group()
    except AssertionError:
        logging.warning('The challenge name in writeup.md (%s) and config.yml (%s) should be the same.' %
                        (writeup_challenge_name, config_challenge_name))
    except Exception:
        logging.warning('The challenge names in writeup.md should be in H1 style. For example:\n\n'
                        'Challenge name\n==============\n')

    # Check if costs and H2 sections are correct in writeup.md
    cost_pattern = '\nCost: [0-9]{1,2}%\n'
    costs = [int(re.search(r'[0-9]{1,2}', cost).group())
             for cost in re.findall(cost_pattern, writeup)]
    if sum(costs) != COST_SUM:
        counted_error('The sum of costs in writeup.md should be %d%%.\n\tPlease make sure you have the following '
                      'format (take care of the white spaces, starting and ending newlines):\n%s'
                      % (COST_SUM, cost_pattern))

    h2_pattern = r'\n## [A-Z].{5,150}\n'
    h2 = re.findall(h2_pattern, writeup)

    if len(h2) < MIN_WRITEUP_SECTIONS:
        counted_error('There should be at least %d sections in writeup.md' % MIN_WRITEUP_SECTIONS)

    h2_last = h2.pop()
    if h2_last != '\n## Complete solution\n':
        counted_error('The last section should be called "Complete solution" in writeup.md')

    h2costs = [re.search(h2_pattern, h2cost).group()
               for h2cost in re.findall(h2_pattern + cost_pattern, writeup)]

    missing_costs = set(h2) - set(h2costs)
    if len(missing_costs) > 0:
        counted_error('No cost is defined in writeup.md for section(s): %s\n\tPlease make sure you have the following '
                      'format (take care of the starting and ending newlines):\n%s'
                      % (missing_costs, cost_pattern))

    reference_style_links_pattern = r'(\[.*?\]\[.*?\])'
    if re.search(reference_style_links_pattern, writeup) is not None:
        logging.warning('The writeup contains reference style links like [this one][0], which might render incorrectly.\n\t'
                        'Please use inline style ones like [this](http://example.net/)')


def _http_request(url):
    # Check controller's test endpoint.
    curl_cmd = ['curl', '-s']

    if url[-4:] == 'test':
        method = 'GET'
        func = 'Test'
    else:
        method = 'POST'
        func = 'Solution checker'
        curl_cmd += ['-X POST']

    curl_cmd += [url]

    try:
        output = subprocess.check_output(curl_cmd).decode('utf-8')
        logging.info('%s output: \n\n%s' % (func, output))

        # Check if solution checker returns a well-formatted JSON response
        solved = False
        if func == 'Solution checker':
            response = json.loads(output)
            solved = bool(response['solved'])
            assert isinstance(response['message'], str)

    except (OSError, subprocess.CalledProcessError) as e:
        # Do not abort if the challenge is not running. TODO: Improve this check!
        logging.warning('Curl returned with error. \n\tDetails: %s' % e)

    except ValueError as e:
        counted_error('Could not parse the response of solution checker. Bad JSON format. \n\tDetails: %s' % e)

    except NameError as e:
        counted_error('JSON attribute "solved" must be a boolean value. \n\tDetails: %s' % e)

    except KeyError as e:
        if e.args[0] == 'solved':
            counted_error('JSON key "%s" is missing from solution checker response. ' % e.args[0])
        elif e.args[0] == 'message':
            logging.warning('JSON key "%s" is missing from solution checker response. \n\t  In certain cases '
                            '(e.g., programming challenges) messages help users to complete the challenge.' % e.args[0])
    except AssertionError:
        counted_error('JSON attribute "message" must be a string value.')

    else:

        if (re.search('OK', output) and func == 'Test') or solved:
            logging.info('%s endpoint passed!' % func)
        elif re.search('404', output) and func == 'Test':
            logging.warning('Test endpoint is not found. '
                            '\n\t  Please implement it so as to make sure your challenge is working correctly.')
        elif re.search('405', output):
            counted_error('%s endpoint should implement HTTP %s.' % (func, method))
        elif re.search('500', output) or not solved:
            counted_error('%s endpoint failed!' % func)


if __name__ == '__main__':
    """
    Sanity check and test an avatao challenge repository. Simply add the challenge repository path as
    the first argument and the script does the rest.

    Python dependencies:
        - PyYAML (http://pyyaml.org/) or simply `pip3 install PyYAML`
          (on Ubuntu you additionally need `apt-get install python3-yaml`)
    """
    init_logger()
    repo_path, repo_name = get_sys_args()
    os.chdir(repo_path)
    sanity_check()
    at_exit()
